Supporting api key from the headers as well

paneru-rajan opened the pull request:
## Supporting api key from the headers as well

<!-- e.g. "Implement user authentication feature" -->

## Relevant issues

Fixes #3332 

## Type
🐛 Bug Fix

## Changes

1. Read the header for `X-API-KEY` if found update the data

I am expecting the feedback, then i will update the PR accordingly.

paneru-rajan committed:
supporting api key from the headers as well

ishaan-jaff commented:
@danenania would this work for your use case ?

ishaan-jaff reviewed (changes requested):
left a review with feedback

paneru-rajan committed:
Support for API Key from  OpenAI / Anthropic Headers

User can set use_llm_key_in_header as true to enable to read the api key from the header
Or, if the api_key is not set, then it tries to get it from header

paneru-rajan commented on litellm/proxy/proxy_server.py:
```diff
@@ -3496,6 +3496,10 @@ async def chat_completion(
             data["api_version"] = query_params["api-version"]
 
+        # Reading API KEY From header
+        if req_header_api_key := request.headers.get('X-API-KEY'):
```
Thank you for the feedback, I've made the fix and added a full description on the main ticket:

https://github.com/BerriAI/litellm/issues/3332#issuecomment-2094819908

paneru-rajan commented:
### Tested for both Anthropic and OpenAI

```
POST http://0.0.0.0:4000/chat/completions
X-API-KEY: sk-ant-api03-.......
Content-Type: application/json

{
  "model": "anthropic/claude-3-opus-20240229",
  "messages": [
    {
      "role": "user",
      "content": "Say this is a test!"
    }
  ],
  "temperature": 0.7
}
```

------

```python
from openai import OpenAI

client = OpenAI(api_key='sk-proj-....', base_url='http://0.0.0.0:4000')

stream = client.chat.completions.create(
    model="gpt-3.5-turbo",
    messages=[{"role": "user", "content": "Say this is a test"}]
    ...
```

### Not Tested!
I am not quite sure, how i can test `litellm.use_llm_key_in_header=True` & all of these again

paneru-rajan commented:
@ishaan-jaff let me know if there's anything i need to do on this, I've accommodated your change request, and tested based on my knowledge.

CC: @krrishdholakia

danenania commented:
I no longer have a direct need for this but will definitely check it out in the future.

krrishdholakia commented on litellm/proxy/proxy_server.py:
```diff
@@ -3496,6 +3496,13 @@ async def chat_completion(
+        if litellm.use_llm_key_in_header or 'api_key' not in data:
+            if auth_bearer := request.headers.get('AUTHORIZATION'):  # OpenAI
+                data['api_key'] = _get_bearer_token(auth_bearer)
+            elif value_of_x_api_key := request.headers.get('X-API-KEY'):  # Anthropic
```
this endpoint is in the openai spec - can you remove this snippet. 

We expect the auth header to be in the bearer token

paneru-rajan replied:
Thank you for the feedback, removed the `'X-API-KEY'` section.

> Also in few line pycharm removed trailing space, i kept that in the commit.

krrishdholakia commented:
Re: testing, 

can you please add a test here - https://github.com/BerriAI/litellm/blob/3be5b72b43ce05851aa34ba05744ea95d1afce70/litellm/tests/test_proxy_server.py#L4

and share a screenshot of it passing 

```
cd litellm/litellm/tests
pytest test_proxy_server.py -x -vv 
```

paneru-rajan committed:
Only used openai spec header

* Removed x-api-key header as we are only supporting openAI specs
* Adjusting the doc-string crawl

paneru-rajan commented:
> Re: testing,
> 
> can you please add a test here -
> 
> https://github.com/BerriAI/litellm/blob/3be5b72b43ce05851aa34ba05744ea95d1afce70/litellm/tests/test_proxy_server.py#L4
> 
> and share a screenshot of it passing
> 
> ```
> cd litellm/litellm/tests
> pytest test_proxy_server.py -x -vv 
> ```

Hi @krrishdholakia 

I tried adding a test case and got stuk:

I was hoping to be able to get the response like below, however it resulted in different error!
```python
response = client.post("/chat/completions", json=test_data, headers= {"Authorization": f"Bearer ...."})

# Exception occurs as soon as i use the Authorization header, possible conflict with the fast-api's Auth header
openai.APIConnectionError: Connection error.
```

Even if this gives us 200 from this test case by altering the mock, we will not be able to test properly as the response will not have the api_key it is using to validate in the test.

Let me know how should i proceed here, already spent quite a lot time figuring things out, your help will be appreciated.


Here is the image of how i ran the test beside what i said here: https://github.com/BerriAI/litellm/pull/3418#issuecomment-2094826655

<img width="574" alt="image" src="https://github.com/BerriAI/litellm/assets/4735661/ce75f0b2-fba2-4336-83b1-7cefb0576835">

paneru-rajan commented:
**Existing test being passed:**

<img width="1254" alt="image" src="https://github.com/BerriAI/litellm/assets/4735661/3b995300-1837-4981-88e8-d1011bddcb51">

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=3418) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=3418) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=3418) it.</sub>
